### PR TITLE
Fixes 3655: Add api to list template rpms

### PR DESCRIPTION
**Pull request by xbhouse**
## Summary

- Adds API for listing content template rpms: `/templates/:uuid/rpms` (similar to the list rpms endpoints for repositories and snapshots)
- Depends on [this](https://github.com/content-services/content-sources-backend/pull/659), will rebase once this is merged

## Testing steps

- Add a custom repository and a RH repository and let them snapshot (this will snapshot a small RH repo: `OPTIONS_REPOSITORY_IMPORT_FILTER=small go run cmd/external-repos/main.go import && go run cmd/external-repos/main.go nightly-jobs`)
- Create a template with those repositories, selecting the latest date 
- Test the `/templates/:uuid/rpms` endpoint with the template uuid and you should see the template's rpms (should include rpms from both of the snapshots in the template) in a response similar to this:
```
{
    "data": [
        {
            "name": "ansible",
            "arch": "noarch",
            "version": "2.8.0",
            "release": "1.el8ae",
            "epoch": "0",
            "summary": "SSH-based configuration management, deployment, and task execution system"
        },
        ...
```        

- To test that the correct snapshot's rpms are shown, you can update the custom repository with a different URL and let it snapshot
- Edit the template date to include this latest snapshot and view the rpms again. The rpms list should reflect the latest snapshot's rpms
- Edit the template to remove the custom repository and then view the rpms again, you should only see the rpms from the RH repository's snapshot

## Checklist

- [ ] Tested with snapshotting feature disabled and pulp server URL not configured if appropriate

### CODE REVIEW

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/HMS-3655

File: pkg/dao/rpms.go [hunk 1]
```diff
@@ -620,3 +619,73 @@ func (r *rpmDaoImpl) ListSnapshotErrata(ctx context.Context, orgId string, snaps
...
+func (r *rpmDaoImpl) ListTemplateRpms(ctx context.Context, orgId string, templateUUID string, search string, pageOpts api.PaginationData) ([]api.SnapshotRpm, int, error) {
+	response := []api.SnapshotRpm{}
+	pulpHrefs := []string{}
+
+	snapshotUuids, err := FetchSnapshotsForTemplate(ctx, r.db, orgId, templateUUID)
+	if err != nil {
+		return response, 0, err
+	}
+
+	res := readableSnapshots(r.db.WithContext(ctx), orgId).Where("snapshots.UUID in ?", UuidifyStrings(snapshotUuids)).Pluck("version_href", &pulpHrefs)
```
Comment (jlsherrill) [inline pkg/dao/rpms.go:633]: Alternatively, FetchSnapshotsForTemplate could return Snapshot objects, and then you could just loop through and pull off the version_href.   That would only result in 1 query (but more data returned for that 1 query)

File: pkg/dao/rpms.go [hunk 2]
```diff
@@ -620,3 +619,73 @@ func (r *rpmDaoImpl) ListSnapshotErrata(ctx context.Context, orgId string, snaps
...
+			Version: pkg.Version,
+			Release: pkg.Release,
+			Epoch:   pkg.Epoch,
+			Summary: pkg.Summary,
+		})
+	}
+	return response, total, nil
+}
+
+func FetchSnapshotsForTemplate(ctx context.Context, db *gorm.DB, orgId string, templateUUID string) ([]string, error) {
```
Comment (jlsherrill) [inline pkg/dao/rpms.go:666]: I would just make this a method on the struct like ListTemplateRpms.  it doesn't have to be declared in the interface and could even be private (by making it start with a lowercase letter)

**Comment by jlsherrill** [comment]
oh and this needs a rebase now

**Review by jlsherrill — APPROVED** [review]

**Comment by swadeley** [comment]
Hi

here we can see three fake packages from the custom repo and the ansible package from the ansible RHEL repo:

```
In [4]: app.content_sources.rest_client.templates_api.list_templates()
2024-05-20 14:12:17.937 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=<snip>, params=[]
Out[4]: 
{'data': [{'arch': 'x86_64',
           'date': '2024-05-19T23:00:00Z',
           'description': 'pr-663',
           'name': 'test-template',
           'org_id': '3340851',
           'repository_uuids': ['19756213-a479-429e-adc6-8334a4a7a077',
                                '7b1386dd-06e6-47dd-a1ff-7e73967dd9ee'],
           'uuid': '6a7c931a-be08-4645-89d7-2793e12dd574',
           'version': '8'}],
 'links': {'first': '/api/content-sources/v1/templates/?limit=100&offset=0',
           'last': '/api/content-sources/v1/templates/?limit=100&offset=0'},
 'meta': {'count': 1, 'limit': 100, 'offset': 0}}

In [6]: app.content_sources.rest_client.templates_api.list_template_rpms('6a7c931a-be08-4645-89d7-2793e12dd574',limit=4)
2024-05-20 14:13:36.891 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=<snip>, params=[('limit', 4)]
Out[6]: 
{'data': [{'arch': 'noarch',
           'epoch': '0',
           'name': 'acme-package',
           'release': '1.elfake',
           'summary': 'acme-package package',
           'version': '1.0.1'},
          {'arch': 'noarch',
           'epoch': '0',
           'name': 'acme-package',
           'release': '1.elfake',
           'summary': 'acme-package package',
           'version': '1.0.2'},
          {'arch': 'noarch',
           'epoch': '0',
           'name': 'acme-package',
           'release': '1.elfake',
           'summary': 'acme-package package',
           'version': '1.1.2'},
          {'arch': 'noarch',
           'epoch': '0',
           'name': 'ansible',
           'release': '1.el8ae',
           'summary': 'SSH-based configuration management, deployment, and '
                      'task execution system',
           'version': '2.8.0'}],
 'links': {'first': '/api/content-sources/v1/templates/6a7c931a-be08-4645-89d7-2793e12dd574/rpms?limit=4&offset=0',
           'last': '/api/content-sources/v1/templates/6a7c931a-be08-4645-89d7-2793e12dd574/rpms?limit=4&offset=64',
           'next': '/api/content-sources/v1/templates/6a7c931a-be08-4645-89d7-2793e12dd574/rpms?limit=4&offset=4'},
 'meta': {'count': 65, 'limit': 4, 'offset': 0}}

In [7]: 
```

**Comment by swadeley** [comment]
Hi

I change the custom repo:

`In [7]: app.content_sources.rest_client.templates_api.list_templates()
2024-05-20 14:19:28.943 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=<snip>, params=[]
Out[7]: 
{'data': [{'arch': 'x86_64',
           'date': '2024-05-19T23:00:00Z',
           'description': 'pr-663 edited',
           'name': 'test-template',
           'org_id': '3340851',
           'repository_uuids': ['19756213-a479-429e-adc6-8334a4a7a077',
                                '461c1b4c-89b1-49aa-abcb-feb391007a1d'],
           'uuid': '6a7c931a-be08-4645-89d7-2793e12dd574',
           'version': '8'}],
 'links': {'first': '/api/content-sources/v1/templates/?limit=100&offset=0',
           'last': '/api/content-sources/v1/templates/?limit=100&offset=0'},
 'meta': {'count': 1, 'limit': 100, 'offset': 0}}
`

At the end of the list we see th eone new fake package:

`In [11]: app.content_sources.rest_client.templates_api.list_template_rpms('6a7c931a-be08-4645-89d7-2793e12dd574',limit=4,offset=56)
2024-05-20 14:21:28.771 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=<snip>, params=[('limit', 4), ('offset', 56)]
Out[11]: 
{'data': [{'arch': 'noarch',
           'epoch': '0',
           'name': 'ansible-test',
           'release': '1.el8ae',
           'summary': 'Tool for testing ansible plugin and module code',
           'version': '2.9.9'},
          {'arch': 'noarch',
           'epoch': '0',
           'name': 'cockateel',
           'release': '1',
           'summary': 'A dummy package of cockateel',
           'version': '3.1'},
          {'arch': 'x86_64',
           'epoch': '0',
           'name': 'sshpass',
           'release': '3.el8ae',
           'summary': 'Non-interactive SSH authentication utility',
           'version': '1.06'}],
 'links': {'first': '/api/content-sources/v1/templates/6a7c931a-be08-4645-89d7-2793e12dd574/rpms?limit=4&offset=0',
           'last': '/api/content-sources/v1/templates/6a7c931a-be08-4645-89d7-2793e12dd574/rpms?limit=4&offset=56',
           'prev': '/api/content-sources/v1/templates/6a7c931a-be08-4645-89d7-2793e12dd574/rpms?limit=4&offset=52'},
 'meta': {'count': 59, 'limit': 4, 'offset': 56}}

In [12]: 

`

**Comment by swadeley** [comment]
Hi

removed the custom repo (notice only one uuid):
```
In [12]: app.content_sources.rest_client.templates_api.list_templates()
2024-05-20 14:31:59.599 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=<snip>, params=[]
Out[12]: 
{'data': [{'arch': 'x86_64',
           'date': '2024-05-19T23:00:00Z',
           'description': 'pr-663 edited again',
           'name': 'test-template',
           'org_id': '3340851',
           'repository_uuids': ['19756213-a479-429e-adc6-8334a4a7a077'],
           'uuid': '6a7c931a-be08-4645-89d7-2793e12dd574',
           'version': '8'}],
 'links': {'first': '/api/content-sources/v1/templates/?limit=100&offset=0',
           'last': '/api/content-sources/v1/templates/?limit=100&offset=0'},
 'meta': {'count': 1, 'limit': 100, 'offset': 0}}

In [13]: app.content_sources.rest_client.templates_api.list_template_rpms('6a7c931a-be08-4645-89d7-2793e12dd574',limit=4,offset=56)
2024-05-20 14:32:06.422 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=e<snip>, params=[('limit', 4), ('offset', 56)]
Out[13]: 
{'data': [{'arch': 'noarch',
           'epoch': '0',
           'name': 'ansible-test',
           'release': '1.el8ae',
           'summary': 'Tool for testing ansible plugin and module code',
           'version': '2.9.9'},
          {'arch': 'x86_64',
           'epoch': '0',
           'name': 'sshpass',
           'release': '3.el8ae',
           'summary': 'Non-interactive SSH authentication utility',
           'version': '1.06'}],
 'links': {'first': '/api/content-sources/v1/templates/6a7c931a-be08-4645-89d7-2793e12dd574/rpms?limit=4&offset=0',
           'last': '/api/content-sources/v1/templates/6a7c931a-be08-4645-89d7-2793e12dd574/rpms?limit=4&offset=56',
           'prev': '/api/content-sources/v1/templates/6a7c931a-be08-4645-89d7-2793e12dd574/rpms?limit=4&offset=52'},
 'meta': {'count': 58, 'limit': 4, 'offset': 56}}

In [14]: 
```

Now the 'A dummy package of cockateel' is gone.

**Pull request merged**